zip-0311.rst

```diff
@@ -0,0 +1,111 @@
+::
+
+  ZIP: 0311 
+  Title: FlyClient ZCash SPV
+  Champions: 
+  Status: Draft
+  Category: Consensus
+  Created: 2019-03-30
+  License: MIT 
+
+
+Terminology
+===========
+The key words "MUST", "SHOULD", and "MAY" in this document are to be interpreted as described in RFC 2119. [#RFC2119]_
+
+Light client
+  A client that is not a full participant in the network of Zcash peers. It can send and receive payments, but does not store or validate a copy of the blockchain.
+
+High probability
+  An event occurs with high probability if it occurs with probability 1-O(1/2^λ), where λ is the security parameter.
+
+Negligible probability
+  An event occurs with negligible probability if it occurs with probability O(1/2^λ), where λ is the security parameter.
+
+Merkle mountain range (MMR)
+  A Merkle mountain range (MMR) is binary hash tree that allows for efficient appends of new leaves without changing the value of existing nodes [#PeterTodd]_.
```

Comment:
```suggestion
  A Merkle mountain range (MMR) is binary hash tree that allows for efficient appends of new leaves without changing the value of existing nodes [#MerkleMR]_.
```

Reply:
Having looked at the Merkle mountain range specification, it is very similar to the incremental Merkle tree that we use for Sprout and Sapling. The primary difference appears to be that the root of our incremental Merkle tree is computed by explicitly filling empty nodes with an unusable value (because we can't have the shape of the tree changing inside the circuit), while the MMR root is constructed almost directly from the roots of its subtrees (which is slightly more efficient, but can't be easily represented within a circuit).

We should think about whether we might want to leverage this new tree of block headers in some future circuit. Would there be a benefit in being able to prove in zero knowledge that some past block contained a particular public value, without revealing which block it was? I'm not sure it's useful without having a circuit-friendly block header construction, and we could probably switch from MMR to a circuit-friendly tree in the future if necessary, but if there's a concrete use case that this could service, we might want to consider replacing the MMR with a circuit-friendly tree from the start.

I haven't confirmed yet whether the FlyClient protocol would still work with an incremental Merkle tree instead of MMR, but it looks like the only change would be an efficiency loss due to the number of empty nodes, which are bounded on `log (n)` (because like MMR, an incremental Merkle tree is filled densely from the left-hand side).